### DISCUSSION TITLE
WPF Application takes too much RAM usage than Winform, how can I optimize WPF App's memory usage?

### DISCUSSION THREAD

**Discussion by neville-studio** [dotnet/wpf#9017] [General]
I have used .NET 8.0 to build two blank applications, the one is WPF application, the another one is WinForm, then publish these applications.

The WPF Application takes too much memory, more than 90MB memory used, but the WinForm application only takes 6.0MB - 7.0 MB memory usage.

So I want to ask how to optimize WPF Application's memory usage if I want to develop a tool app.

**Comment by lindexi** [comment]
@neville-studio Given that memory measurement is a complex topic, to avoid any misunderstandings due to differing perceptions, I highly recommend you try out a simple application. For instance, the code I provide below is a WPF window program that launches through the console, located in Program.cs. This is a very straightforward application, please ensure to switch to the x86 platform and run it after releasing it. After running, you can try to observe the size of the memory occupied by this process. To avoid causing you any confusion, I have also uploaded all the code from my project to [GitHub](https://github.com/lindexi/lindexi_gd/tree/8633b2b1db7c87332869ff2f5f89731ce77b5f25/HayceewiballfergeBawbalhega). You can download all the files from my project, open the HayceewiballfergeBawbalhega.sln file in VisualStudio, then switch to Release mode to build and run the application by double-clicking the exe file in Explorer, independent of VisualStudio.

```csharp
using System.Windows;
using System.Windows.Interop;
using System.Windows.Media;

namespace HayceewiballfergeBawbalhega;

class Program
{
    [STAThread]
    static void Main()
    {
        // Enable SoftwareOnly RenderMode to avoid hardware effect.
        RenderOptions.ProcessRenderMode = RenderMode.SoftwareOnly;

        var application = new Application();
        application.Startup += (s, e) =>
        {
            new Window()
            {
                Width = 200,
                Height = 100,
            }.Show();
        };
        application.Run();
    }
}
```

**Reply by neville-studio** [reply]
I have created two C# solutions, the first one is used WPF Application template in Visual Studio 2022. The another one have used Windows Forms App. The two solutions all have not added any code, and then published it. In Task Manager, I have found the WPF release have used more than 90MB RAM, but the Windows Forms application only have used 6.0MB - 7.0 MB RAM. So, I don't know how to optimize this application.

By the way, the project you have provided have used about 12MB RAM.

After I have using following code in `App.xaml.cs` in WPF default template and then publish again. This release have only takes about 20MB RAM usage

```C#
  RenderOptions.ProcessRenderMode = RenderMode.SoftwareOnly;
```

Because I don't know how to optimize a WPF application, and this template has taken too many RAM usage in Windows. And I want make a tool application which consumes less RAM and less CPU usage , So I used back to Windows Forms to make a app.

**Reply by lindexi** [reply]
@neville-studio In fact, if you had an understanding of memory and hardware rendering, you probably wouldn't be asking questions and making technical decisions like this.

The issue you're seeing with WPF consuming a large amount of memory is essentially due to the memory used by the hardware rendering portion. This can be influenced by many factors, including hardware devices, driver software, system versions, and so forth. This is why I insist on you using the SoftwareOnly option. Furthermore, the memory consumption of an application is quite complex. From a professional standpoint, you can't solely rely on the memory indicated in the task manager to understand the memory usage of your application process. Instead, you should use professional tools like vmmap for measurement. The memory consumption of an application is divided into multiple parts and dimensions, including shared and private memory from the perspective of shared and exclusive usage. From another dimension, the application's memory can be partly in the virtual memory on the disk and in the physical memory on the memory stick. As for the Windows Forms portion you mentioned consuming less memory, this is because some of the memory in the System has not been accounted for by you, giving you a false impression. However, I fully support you in using the technology you are most familiar with to complete your tasks.

**Reply by neville-studio** [reply]
I'm just a C# learner, I don't know some details in WPF technology. Anyway, thank you.

I just want to make an application universal and less RAM usage, CPU usage which task manager shows. If users see this application consumes more CPU or memory usage in their task manager, they might don't like this application.

I previously used Win32 technology to develop an app. It consumes less RAM and less CPU usage that task manager shows, but it brings many graphics bug and I don't know how to optimize it, so I migrate to C#. 

If there has any way to make this application consumes less CPU and RAM usage that task manager shows, I will consider using WPF technology to develop it.

I will turn to Windows Forms (WinForm) technology to develop this C# app. it met my expectations.

Thanks your answer!
